### ISSUE TITLE
OSC clip launching and recording issues

### ISSUE THREAD

**Issue by metakeule** [git-moss/DrivenByMoss#93]
Hi,

thank you very much for your effort in making these extensions.
I am trying to remote control bitwig studio via OSC and my monome and
try to provide visual feedback on the monome.

This is my experience so far.

The clips in bitwig have the following states:

1. empty
2. stopped
3. scheduled to play
4. scheduled to record
5. scheduled to stop
6. playing
7. recording

I would like to reliably switch between each of these states via OSC.

So far we have the following commands
```
/refresh
/track/{1-8}/clip/{1-N}/launch
/track/{1-8}/clip/{1-N}/record
/track/{1-8}/clip/{1-N}/select
/track/{1-8}/clip/stop
```

And the following messages send out from bitwig

```
/track/{1-8}/clip/{1-N}/hasContent
/track/{1-8}/clip/{1-N}/isPlaying
/track/{1-8}/clip/{1-N}/isPlayingQueued
/track/{1-8}/clip/{1-N}/isRecording
```

Now there are lots of differences between what I would expect from these commands and responses and what currently happens (bitwig 2.4, OSC 6.0).

## "/track/{1-8}/clip/{1-N}/launch"
**I would expect the following behavior**

If the slot is filled, it would be scheduled to play,
then played, stopping any other clip on the same track.
When it is scheduled, the "isPlayingQueued" message would be
send with a value of 1, when it is playing, the "isPlaying" 
message would be send with a value of 1.

If the slot is not filled, nothing would happen, regardless of
the track being armed or not.

When the clip is going to be played the general sequencer will start playing,
if it is not already.

**The current behavior is the following:**

When the slot is empty and the track is not armed, nothing happens.
However when the track is armed, it starts recording! (Although there is a seperate record command). Since it is neccessary to arm a track in order to play
a virtual instrument one can easily record "by accident".
Furthermore,  "isPlayingQueued" is sent (?) and "isRecording" is not sent, so we
can't give visual feedback on the monome. Instead "isRecording" is sent with a value of 0 when the recording stopped.

## "/track/{1-8}/clip/{1-N}/record"
**I would expect the following behavior**

If the slot is filled, and overdup clip mode is not active, clear the slot and record newly. If overdup clip mode is active, overdup to the existing slot.
If the track is not armed, it should be armed, since we told to record.
A "isRecording" message should be send with a value of 1 confirming the recording.

Recording would start no matter, if the sequencer is playing or not.

**The current behavior is the following:**

If the slot is empty and the sequencer is not playing, nothing happens no matter,
if the track is armed or not. However, if the slot is empty and the sequencer is playing, it will record in the slot, even if the track is not armed (and will not arm the track magically).

If the slot is not empty and the sequencer is running, it depends on wether the 
track is armed or not, if it will record into the slot.

In all cases there is no feedback message confirming the start and end of the recording.

I have not fully understood when it is possible to directly play the clip and have the recording being stopped and when not.

## further notes

Since some of the behavior depends on whether slots are filled or not and wether
tracks are armed or not, it would be good to get this information from bitwig
at the start of a session (e.g. after calling "/refresh") (e.g. via "/track/{1-8}/clip/{1-N}/hasContent").

I don't know, if these are bugs within bitwig, or if you could reliable expose
those states via OSC. That would be great.

**Comment by git-moss** [comment]
You are right the behaviour of launch and record is not consistent. I will check back with Bitwig what the idea is here.
I also checked the state transfers and they look good to me (I used the Max patch coming with the download). Also on /refresh I get all the correct hasContent states.

**Comment by metakeule** [comment]
For now, the most reliable way is just to launch no matter what and live with the fact, that when there is a clip, it will be played and otherwise recorded. However since track arming is also not reliable (see #95) and launching only triggers recording, if the track is armed, all tracks always have to be armed... ;-)

I'll check the transfer with the Max patch.

**Comment by metakeule** [comment]
When sending "/refresh" without a value, I get 

```
/track/8/clip/1/isSelected [0]
bitwig said: /browser/filter/4/item/14/exists [0]
```

while having 15 slots spread over 4 tracks filled with a clip.
I am running bitwig on linux (ubuntu) in case it makes a difference.

**Comment by git-moss** [comment]
What client or OSC library are you using? Could it be that you only read the first message of a Bundle?

**Comment by metakeule** [comment]
It could be. I am using github.com/metakeule/osc which is a slightly (only for debugging) modified fork of https://github.com/scgolang/osc .

**Comment by git-moss** [comment]
Sending of bundles is synchronized, so this should not happen.
